src/transformers/modeling_rag.py

```diff
@@ -0,0 +1,950 @@
+
+        retriever_inputs = self.question_encoder_tokenizer.batch_encode_plus(
+            input_strings,
+            return_tensors="pt",
+            padding=True,
+            truncation=True,
+        )
+        retriever_input_embs = self.question_encoder(retriever_inputs["input_ids"].to(device))[0]
+
+        doc_scores, docs = self.retriever.retrieve(retriever_input_embs, n_docs=self.n_docs)
+
+        rag_input_strings = [
+            self._cat_input_and_doc(
+                doc_scores[i][j], docs[i]["title"][j], docs[i]["text"][j], input_strings[i], add_eos, print_docs
+            )
+            for i in range(len(docs))
+            for j in range(self.n_docs)
+        ]
+
+        contextualized_inputs = self.generator_tokenizer.batch_encode_plus(
+            rag_input_strings,
+            max_length=self.config.max_combined_length,
+            return_tensors="pt",
+            padding="max_length",
+            truncation=True,
+        ).to(device)
+
+        return (
+            contextualized_inputs["input_ids"],
+            contextualized_inputs["attention_mask"],
+            doc_scores.to(device),
+        )
+
+
+class RAGEncoder(torch.nn.Module):
+    r"""
+    RAG is an encoder-decoder model, however, we don't exaplicitly implement an encoder and a decoder layes,
+    like it's done e.g. in BART and T5 implementations - for RAG these are encapsulated inside the generaotr instance.
+    This is a dummy model simulating RAG encoder output, need for compatibility with transformers generation code
+    """
+
+    def __init__(self, rag_model: RagModel):
+        super().__init__()
+        self.rag_model = rag_model
+
+    def forward(self, input_ids=None, attention_mask=None):
+        ctxt_input_ids, ctxt_attention_mask, doc_scores = self.rag_model.contextualize(input_ids, print_docs=False)
+        encoder = self.rag_model.generator.get_encoder()  # stacked
+        encoder_outputs = encoder(input_ids=ctxt_input_ids, attention_mask=ctxt_attention_mask, return_dict=True)
+        # needed to satisfy assertions in generation_utils
+        unstacked_x = _unstack_ctxt(encoder_outputs.last_hidden_state, self.rag_model.n_docs)
+
+        return (unstacked_x, encoder_outputs.hidden_states, ctxt_attention_mask, doc_scores)
+
+
+class PreTrainedRagModel(PreTrainedModel):
+    r"""
+    RAG models encapsulate two trainable components - a question encoder and a generator, but as such they don't have any trainable parameters.
+    We specialize `:func:`~transformers.PreTrainedModel.from_pretrained`` and `:func:`~transformers.PreTrainedModel.save_pretrained`` to reflect this.
+    """
+    config_class = RagConfig
+
+    def __init__(
+        self,
+        config: RagConfig,
+        retriever,
+        question_encoder,
+        question_encoder_tokenizer,
+        generator,
+        generator_tokenizer,
+    ):
+        super().__init__(config)
+        self.model = RagModel(
```

Comment:
I think we can adopt the approach of our `EncoderDecoderModel` here by having the `RAGModel` inherit from the `PreTrainedRagModel` class instead of encapsulating it this will remove a level in the hierarchy.

Also, the `EncoderDecoderModel` class was designed to let you instantiate a model along two possible paths:
- load from two separate checkpoints, one for the encoder, one for the decoder
- load from a single seq2seq model comprising both components.

Maybe we want to adopte the same approach here for flexibility.